apps/arweave/src/ar_mining_worker.erl

```diff
@@ -611,257 +693,187 @@ passes_diff_checks(SolutionHash, IsPoA1, Candidate, State) ->
+			report_chunk_cache_metrics(State1),
+			{true, State1};
+		{error, Reason} ->
+			?LOG_WARNING([{event, mining_worker_failed_to_reserve_cache_space},
+				{worker, State#state.name}, {partition, State#state.partition_number},
+				{session_key, ar_nonce_limiter:encode_session_key(SessionKey)},
+				{reason, Reason}]),
 			false
 	end.
 
-do_not_cache(Candidate, State) ->
+mark_chunk2_missing(Candidate, State) ->
 	#mining_candidate{ packing_difficulty = PackingDifficulty } = Candidate,
-	do_not_cache(0, ar_block:get_max_nonce(PackingDifficulty), Candidate, State).
+	mark_chunk2_missing(0, ar_block:get_max_nonce(PackingDifficulty), Candidate, State).
 
-do_not_cache(Nonce, NonceMax, _Candidate, State)
-		when Nonce > NonceMax ->
+mark_chunk2_missing(Nonce, SubchunksPerRecallRange, _Candidate, State)
+when Nonce > SubchunksPerRecallRange ->
 	State;
-do_not_cache(Nonce, NonceMax, Candidate, State) ->
-	State2 = cache_chunk(do_not_cache, Candidate#mining_candidate{ nonce = Nonce }, State),
-	do_not_cache(Nonce + 1, NonceMax, Candidate, State2).
-
-%% @doc The sub_chunk_cache stores either the first or second chunk for a given nonce. This is
-%% because we process both the first and second recall ranges in parallel and don't know
-%% which data will be available first. The function manages that shared cache slot by either
-%% caching data if its the first to arrive, or "popping" data that was previously cached. The
-%% caller is responsible for taking the appropriate action based on the return value.
-%%
-%% do_not_cache is a special value used to prevent unnecessary data from being cached once a
-%% solution has been found for a given nonce.
-cycle_sub_chunk_cache(#mining_candidate{ cache_ref = CacheRef } = Candidate, Data, State)
-  		when CacheRef /= not_set ->
-	#mining_candidate{ nonce = Nonce, session_key = SessionKey } = Candidate,
-	Cache = State#state.sub_chunk_cache,
-	SessionCache = maps:get(SessionKey, Cache, #{}),
-	case maps:take({CacheRef, Nonce}, SessionCache) of
-		{do_not_cache, SessionCache2} ->
-			Cache2 = maps:put(SessionKey, SessionCache2, Cache),
-			{do_not_cache, State#state{ sub_chunk_cache = Cache2 }};
-		error ->
-			{cached, cache_chunk(Data, Candidate, State)};
-		{CachedData, SessionCache2} ->
-			Cache2 = maps:put(SessionKey, SessionCache2, Cache),
-			{CachedData, State#state{ sub_chunk_cache = Cache2 }}
+mark_chunk2_missing(Nonce, SubchunksPerRecallRange, Candidate, State) ->
+	case ar_mining_cache:with_cached_value(
+		?VDF_STEP_CACHE_KEY(Candidate#mining_candidate.cache_ref, Nonce),
+		Candidate#mining_candidate.session_key,
+		State#state.chunk_cache,
+		fun(CachedValue) -> {ok, CachedValue#ar_mining_cache_value{chunk2_missing = true}} end
+	) of
+		{ok, ChunkCache1} ->
+			mark_chunk2_missing(Nonce + 1, SubchunksPerRecallRange, Candidate, State#state{ chunk_cache = ChunkCache1 });
+		{error, Reason} ->
+			%% NB: this clause may cause a memory leak, because mining worker will wait for
```

Comment:
Given the function used here, there's no way this clause can be hit, right? i.e. 

```
fun(CachedValue) -> {ok, CachedValue#ar_mining_cache_value{chunk2_missing = true}} end
```

never returns {error, Reason}

If so, maybe we drop the clause and comment? Mostly just thinking of future engineers going through the code looking for memory leaks

Reply:
There is still a way to get an error here, for example, if the session was removed (`{error, session_not_found}`).